### PR TITLE
Experiment with CircleCI "resource_class"

### PR DESCRIPTION
**Pull request by kripken**
This sees what happens if we use "extra large" - the biggest CircleCI has - compared
to the default "medium" we've been using. This should bump us from 2 virtual cores to
8 and likewise increase memory by a similar ratio.

Sadly this does not help on `build-linux` that I can see, which is the most critical one.
But it does help on `core*` and `other`, but only by around 30% or so, which is slightly
disappointing given how parallelizable our tests are (and even inside each test, in
binaryen). Perhaps disk I/O or something else is just significant there?

I can't find any docs on the cost of this switch, but I guess their pricing may not be
fully public? Overall, given the modest improvement here, I'm not sure it's worth it,
but if the cost is not an issue maybe we should do it.

### CODE REVIEW

**Comment by dschuff** [comment]
It seems quite surprising that going from 2 cores to 4 or 8 doesn't help with e.g. the embuilder ALL stage. Even individual builds like libc are no faster. On my workstation the libc build is almost 2x as fast going from 2 cores to 4. I wonder if the config isn't quite right, or if this is somehow I/O bound?

**Comment by sbc100** [comment]
I think you want to bump (or just remove) the `EMCC_CORES` env var setting at the top of this file.

**Comment by kripken** [comment]
Good catch @sbc100 ! That seems most likely. I'll push a few options to test values there.

**Comment by dschuff** [comment]
wow I didn't notice that either. Is there a particular reason it needs to be there? do the scripts not reliably detect the number of cores on these VMs?

**Comment by dschuff** [comment]
(after the 8-core run, can you maybe try the 4-core one too, for completeness?)

**Comment by sbc100** [comment]
> wow I didn't notice that either. Is there a particular reason it needs to be there? do the scripts not reliably detect the number of cores on these VMs?

I think I wanted to spawn more jobs than CPUs .. normally we just do 1 job per CPU, but for I/O bound things its often better to do 2x.   I think we are trying to do 2x the number cores here, and I guess we saw a speedup.

**Comment by sbc100** [comment]
(So you might want to try with 16)

**Comment by kripken** [comment]
Still experimenting with values, but now we are in the right place, as doubling the cores from 4 to 8 almost halved the time.

**Comment by kripken** [comment]
Here is some data:
build-linux, core0, core3, other and browser-chrome
There is significant noise (see first 2 column groups) but it does seem like more cores help, and x-large is better than large which is better than medium.

I'm leaning towards x-large without specifying the cores, because the autodetection seems good enough, and is more future proof. Thoughts?

The overall win here to total iteration time is limited, however, by the speed of `browser-chrome` which only improves a little - we use little parallelism there. But we could perhaps split that into two jobs.

**Comment by sbc100** [comment]
Yes x-large without explicit NUM_CORES sounds good to me!

Thanks for taking the time!

**Review by sbc100 — APPROVED** [review]

**Comment by sbc100** [comment]
(BTW do we know the cost implications of switching to x-large?)

**Comment by kripken** [comment]
Ok, this is ready for review now I think. It only adjusts `build-linux` to use `x-large`, and with 16 cores. I avoided changing anything else to avoid confusion in any future bisections or regressions that this causes.

The benefit of the change is a 2x speedup in `build-linux`, from ~50 minutes to ~25.

File: .circleci/config.yml [hunk 1]
```diff
@@ -303,6 +303,13 @@ jobs:
           test_targets: "sanity"
   build-linux:
     executor: bionic
+    # xlarge has 4x the cores of the default medium, costs 4x as much, and runs
+    # in about 1/2 the time, so it is not cost-effective (overall it is 2x the
+    # cost for the same work), but given this blocks almost all the other jobs
+    # we want it to finish asap
+    resource_class: xlarge
+    environment:
+      EMCC_CORES: 16
```
Comment (sbc100) [inline .circleci/config.yml:312]: Why not just remove EMCC_CORES from here and above?  (Or do you want to do that as a separate change?)

**Review by sbc100 — APPROVED** [review]

File: .circleci/config.yml (hunk 1 above)
Comment (kripken) [reply .circleci/config.yml:312]: Yes, I think it's best to do any other changes separately.

**Review by dschuff — APPROVED** [review]

**Pull request merged**